File: Leetcode_Array/src/Array/A08_Max_Consecutive_Ones.java
package Array;
import java.lang.reflect.Array;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by happiness on 17/10/13.
 * Given a binary array, find the maximum number of consecutive 1s in this array.
 * <p/>
 * Example 1:
 * Input: [1,1,0,1,1,1]
 * Output: 3
 * Explanation: The first two digits or the last three digits are consecutive 1s.
 * The maximum number of consecutive 1s is 3.
 * Note:
 * <p/>
 * The input array will only contain 0 and 1.
 * The length of input array is a positive integer and will not exceed 10,000
 */
public class A08_Max_Consecutive_Ones {
    public static void main(String args[]) {

        int[] nums = {1, 3, 1, 15};
        System.out.println((findMaxConsecutiveOnes(nums)));

    }


    public static int ss_findMaxConsecutiveOnes(int[] nums) {
        int res = 0;
        int count = 0;
        for (int num : nums) {
            if (num == 1) {
                count++;
            } else {
                count = 0;
            }
            res = Math.max(res, count);
        }

        return res;
    }

    public static int findMaxConsecutiveOnes(int[] nums) {


        int now = 0;
        int max = 0;
        for (int n : nums) {
            // max = Math.max(max, now = n == 0 ? 0 : ++now); // ++NOW 先赋值，再使用，可以是 now + 1，不能是i++,


            max = Math.max(max, now = (n == 1 ? now + 1 : 0));

//            三元复杂表达式，还是用N + 1，避免使用N++或者++N
//            max = Math.max(max, now = (n == 1 ? ++now : 0));


            /*  解释：
            if (n == 1) now++;
            else now = 0;
            max = Math.max(max, now);



            if (n == 1) {
                now++;
            } else {
                now = 0;
            }
            max = Math.max(max, now);
            */
        }
        return max;

    }

}

File: Python/A08_Max_Consecutive_Ones.py
'''
Given a binary array, find the maximum number of consecutive 1s in this array.

Example 1:
Input: [1,1,0,1,1,1]
Output: 3
Explanation: The first two digits or the last three digits are consecutive 1s.
    The maximum number of consecutive 1s is 3.
Note:
'''
class Solution:
    def findMaxConsecutiveOnes(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        now = 0;res = 0
        for num in nums:
        	if num == 1:
        		now += 1
        	else:
        		now = 0
        	res = max(now,res)
        return res

    def findMaxConsecutiveOnes_2(self, nums):
    	'''
    	思路：(''.join(map(str,nums))) 整形列表转字符串
    	字符串用.split函数分割成数组

    	或者
        return max(map(lambda x: len(x), ''.join([str(num) for num in nums]).split('0')))

    	'''
        a = (''.join(map(str,nums))).split('0')
        return max(map(len,a))


    def findMaxConsecutiveOnes_3(self, nums):
    	'''
		transform the input, like
		[0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1], to
		[0, 1, 2, 3, 0, 1, 2, 0, 0, 1, 2, 3, 4], and the max element would be the result.
    	'''
        for i in range(1, len(nums)):
            if nums[i]:
                nums[i] += nums[i - 1]
        return max(nums)




so = Solution()
nums = [1,1,0,1,1,1]
print(so.findMaxConsecutiveOnes(nums))